File: game_of_life.rb
# @param {Integer[][]} board
# @return {Void} Do not return anything, modify board in-place instead.
def game_of_life(board)
  # clone and clear board
  clone = Marshal.load(Marshal.dump(board))
  ymax = clone.length
  xmax = clone[0].length

  board.each_with_index do |y_row, yid|
    y_row.each_with_index do |x_row, xid|
      board[yid][xid] = calc_life(clone, yid, xid, ymax, xmax)
    end
  end
end

def calc_life(clone, yid, xid, ymax, xmax)
  sum, ans = 0, 0
  [-1,0,1].each do |y|
    [-1,0,1].each do |x|
      unless y == 0 && x == 0
        ystep = yid + y
        xstep = xid + x
        sum += (clone[ystep][xstep] || 0) if within_limit(ystep, xstep, ymax, xmax)
      end
    end
  end

  ans =
  if clone[yid][xid] == 1
    # Any live cell with fewer than two live neighbors dies, as if caused by under-population.
    if sum < 2
      0
    # Any live cell with two or three live neighbors lives on to the next generation.
    elsif sum == 2 || sum == 3
      1
    # Any live cell with more than three live neighbors dies, as if by over-population..
    else
      0
    end
  else
    # Any dead cell with exactly three live neighbors becomes a live cell, as if by reproduction.
    sum == 3 ? 1 : 0
  end

  ans
end

def within_limit(ystep, xstep, ymax, xmax)
  ystep < ymax && ystep >=0 && xstep < xmax && xstep >= 0 ? true : false
end

File: game-of-life.js
/**
 * @param {number[][]} board
 * @return {void} Do not return anything, modify board in-place instead.
 */
// O(n*m)
// S(1)

const count = (row, col, board) => {
    let total = 0;
    let directions = [[-1,-1],[-1,0],[-1,1],[0,-1],[0,1],[1,-1],[1,0],[1,1]];
    for (let dir=0; dir<directions.length; dir++) {
        let d=directions[dir];
        if (board[row+d[0]] && Math.abs(board[row+d[0]][col+d[1]]) === 1) total++;
    }
    return total;
}

const gameOfLife = (board) => {
    for (let r=0; r<board.length; r++) {
        for (let c=0; c<board[0].length; c++) {
            let neighbors = count(r,c,board);
            if (board[r][c]===1 || board[r][c]===-1) {
                // Any live cell with fewer than two live neighbors dies
                // Any live cell with more than three live neighbors dies
                if (neighbors < 2 || neighbors > 3) board[r][c] = -1;
            } else {
                // Any dead cell with exactly three live neighbors becomes a live cell
                if (neighbors===3) board[r][c] = 2;
            }
        }
    }
        
    for (let i=0; i<board.length; i++) {
        for (let j=0; j<board[i].length; j++) {
            if (board[i][j]<0) board[i][j] = 0;
            if (board[i][j]===2) board[i][j] = 1;
        }
    }
    return board;
};
